runtime/compiler/ilgen/Walker.cpp

```diff
@@ -1688,6 +1697,8 @@ TR_J9ByteCodeIlGenerator::stashArgumentsForOSR(TR_J9ByteCode byteCode)
       numArgs -= 1;
       if (unresolvedInCP)
          numArgs -= 1;
+      else if (isInvokeCacheAppendixNull)
+         numArgs += 1;
```

Comment:
I don't think I understand why `numArgs` should be adjusted in this case. Won't we be passing precisely the number and type of arguments expected by the callee, i.e. expected by the signature of `symRef`?

Reply:
This `numArgs` adjustment is to get its value back to what the callee signature expects in case appendix is null. Perhaps a clearer way to put this is:
```
if (!isInvokeCacheAppendixNull)
   {
   numArgs -=1;
   if (unresolvedInCP)
      numArgs -= 1;
   }
```
What do you think?

Reply:
Oh, I guess we're trying to get back to the number of arguments expected by the signature in the bytecode, which doesn't include any appendix. That explains what the first `numArgs -= 1` is doing there in the first place. Then your added code is undoing the appendix adjustment

If that's right, I think your proposed change is a lot clearer. Maybe even better with a few short comments, e.g.
```
if (!isInvokeCacheAppendixNull)
   {
   // Only count arguments that are on the stack before the invoke bytecode instruction
   numArgs -= 1; // appendix
   if (unresolvedInCP)
      numArgs -= 1; // MemberName
   }
```
Though understanding this adjustment (again, if the above is right) also makes me suspicious that when there is an appendix we might not be storing the right argument values. For example, suppose the signature is `(I)`, and suppose there's an appendix, so the callee expects an integer and a reference, say `(ILjava/lang/Object;)`. Before the call, the stack looks like `... n`, where `n` is the int argument. Then for the purpose of the call we also push `appendix`, so the stack is now `... n appendix`. It looks like the code just below will treat the top `numArgs` stack values as the arguments, so when we reduce `numArgs` from 2 to 1, I think it will pick up `appendix`, but if it's supposed to find the arguments according the original signature `(I)`, then it should pick up `n` instead

Reply:
The new changes in `genInvokeDynamic` and `genInvokeHandle` would ensure we do not perform a push of the appendix object when it is NULL. `loadInvokeCacheArrayElements` is called only if appendix object is not null, or if the invokedynamic/handle is unresolved.

Reply:
I think that's consistent with the understanding I described in my last message. The possible problem I mentioned is in the (existing) case where there *is* a (non-null) appendix.

Reply:
Oh, I misunderstood what you were describing initially. I also see that the code below would end up picking up the appendix object as the last arg anyway. I guess we could change usage of `_stack.size()` in the conditions to some variable set to `_stack.size() - numObjectsToNotStash` to actually ignore the appendix object.